olive/passes/onnx/bnb_quantization.py

```diff
@@ -0,0 +1,124 @@
+            quantized_modules = quantized_modules or model.model_attributes.get("quantized_modules")
+
+            # extract quant_type from model_attributes if not specified in config
+            if not quant_type:
+                quant_config = model.model_attributes.get("quantization_config")
+
+                if quant_config.get("load_in_8bit"):
+                    raise ValueError("load_in_8bit is not supported. Only 4-bit quantization is supported.")
+
+                quant_type = quant_config.get("bnb_4bit_quant_type")
+                if not quant_type:
+                    raise ValueError("quant_type is required.")
+
+                if quant_config.get("bnb_4bit_use_double_quant"):
+                    logger.info(
+                        "bnb_4bit_use_double_quant is set to True but double quantization is not supported. Ignoring."
+                    )
+        assert quant_type in ["fp4", "nf4"], f"quant_type must be one of 'fp4' or 'nf4'. Got {quant_type}."
+        quant_type_enum = getattr(MatMulBnb4Quantizer, quant_type.upper())
+
+        # load the model
+        onnx_model = model.load_model()
+
+        # get nodes to exclude from quantization
+        nodes_to_exclude = config["nodes_to_exclude"] or []
+
+        # find all MatMul nodes in the graph
+        matmul_nodes = self._find_matmul_nodes(onnx_model.graph)
+        # filter based on quantized_modules
+        quantized_modules = set(quantized_modules or [])
+        nodes_to_exclude = nodes_to_exclude + [
+            node
+            for node in matmul_nodes
+            if quantized_modules and not any(re.match(f".*[./]{key}[./]MatMul$", node) for key in quantized_modules)
+        ]
+
+        # quantize the model
+        quantizer = MatMulBnb4Quantizer(
+            onnx_model, quant_type=quant_type_enum, block_size=64, nodes_to_exclude=nodes_to_exclude
+        )
+        quantizer.process()
+        # topologically sort the graph at the end since previous optimizations may have broken it
+        quantizer.model.topological_sort()
+
+        # save the model to the output path and return the model
+        return model_proto_to_olive_model(onnx_model, output_model_path, config)
+
+    def _find_matmul_nodes(self, graph: onnx.GraphProto) -> List[str]:
+        """Find all MatMul nodes in the graph and return their names."""
+        matmul_nodes = []
+        for node in graph.node:
+            for attr in node.attribute:
+                if attr.type == onnx.AttributeProto.GRAPH:
+                    # recursive call to take care of sub-graph
+                    matmul_nodes += self._find_matmul_nodes(attr.g)
+                elif attr.type == onnx.AttributeProto.GRAPHS:
+                    for subgraph in attr.graphs:
+                        # recursive call to take care of sub-graph
+                        matmul_nodes += self._find_matmul_nodes(subgraph)
+            if node.op_type == "MatMul":
+                matmul_nodes.append(node.name)
```

Comment:
yield has better performance than loop.

Reply:
Will address this in a follow up PR as discussed offline